File: implement-shell-tools/wc/wc.py
#!/usr/bin/env python3
import sys


def count_bytes(content):
    lines = content.count(b'\n')
    words = len(content.split())
    chars = len(content)
    return lines, words, chars


def print_row(lines, words, chars, label, flag_l, flag_w, flag_c, no_flag):
    suffix = f" {label}" if label else ""
    if no_flag:
        print(f"{lines:8}{words:8}{chars:8}{suffix}")
    elif flag_l:
        print(f"{lines:8}{suffix}")
    elif flag_w:
        print(f"{words:8}{suffix}")
    elif flag_c:
        print(f"{chars:8}{suffix}")


def main():
    args = sys.argv[1:]
    flag_l = False
    flag_w = False
    flag_c = False
    files = []

    for arg in args:
        if arg == '-l':
            flag_l = True
        elif arg == '-w':
            flag_w = True
        elif arg == '-c':
            flag_c = True
        else:
            files.append(arg)

    no_flag = not (flag_l or flag_w or flag_c)
    exit_code = 0

    if not files:
        content = sys.stdin.buffer.read()
        lines, words, chars = count_bytes(content)
        print_row(lines, words, chars, '', flag_l, flag_w, flag_c, no_flag)
    else:
        total_lines = total_words = total_chars = 0
        counted = []
        for filepath in files:
            try:
                with open(filepath, 'rb') as f:
                    content = f.read()
                lines, words, chars = count_bytes(content)
                total_lines += lines
                total_words += words
                total_chars += chars
                counted.append((lines, words, chars, filepath))
                print_row(lines, words, chars, filepath, flag_l, flag_w, flag_c, no_flag)
            except OSError as e:
                print(f"wc: {filepath}: open: {e.strerror}", file=sys.stderr)
                exit_code = 1

        if len(files) > 1:
            print_row(total_lines, total_words, total_chars, 'total',
                      flag_l, flag_w, flag_c, no_flag)

    sys.exit(exit_code)


if __name__ == '__main__':
    main()

File: implement-shell-tools/wc/wc.js
#!/usr/bin/env node
// Usage: node wc.js [-l] [-w] [-c] file1 [file2 ...]
const fs = require('fs');

const args = process.argv.slice(2);
let flagL = false;
let flagW = false;
let flagC = false;
const files = [];

for (const arg of args) {
  if (arg === '-l') flagL = true;
  else if (arg === '-w') flagW = true;
  else if (arg === '-c') flagC = true;
  else files.push(arg);
}

const noFlag = !(flagL || flagW || flagC);

function osError(err) {
  if (err.code === 'EACCES') return 'Permission denied';
  if (err.code === 'ENOENT') return 'No such file or directory';
  return err.message;
}

function countBytes(buf) {
  const lines = Array.from(buf).filter(b => b === 10).length;
  const text = buf.toString('utf8').trim();
  const words = text === '' ? 0 : text.split(/\s+/).length;
  const chars = buf.length;
  return { lines, words, chars };
}

function countString(str) {
  const lines = (str.match(/\n/g) || []).length;
  const text = str.trim();
  const words = text === '' ? 0 : text.split(/\s+/).length;
  const chars = str.length;
  return { lines, words, chars };
}

function printRow(lines, words, chars, label) {
  const suffix = label ? ` ${label}` : '';
  if (noFlag) {
    process.stdout.write(
      `${String(lines).padStart(8)}${String(words).padStart(8)}${String(chars).padStart(8)}${suffix}\n`
    );
  } else if (flagL) {
    process.stdout.write(`${String(lines).padStart(8)}${suffix}\n`);
  } else if (flagW) {
    process.stdout.write(`${String(words).padStart(8)}${suffix}\n`);
  } else if (flagC) {
    process.stdout.write(`${String(chars).padStart(8)}${suffix}\n`);
  }
}

if (files.length === 0) {
  let content = '';
  process.stdin.setEncoding('utf8');
  process.stdin.on('data', chunk => { content += chunk; });
  process.stdin.on('end', () => {
    const { lines, words, chars } = countString(content);
    printRow(lines, words, chars, '');
  });
} else {
  let exitCode = 0;
  let totalLines = 0, totalWords = 0, totalChars = 0;

  for (const filepath of files) {
    try {
      const buf = fs.readFileSync(filepath);
      const { lines, words, chars } = countBytes(buf);
      totalLines += lines;
      totalWords += words;
      totalChars += chars;
      printRow(lines, words, chars, filepath);
    } catch (err) {
      process.stderr.write(`wc: ${filepath}: open: ${osError(err)}\n`);
      exitCode = 1;
    }
  }

  if (files.length > 1) printRow(totalLines, totalWords, totalChars, 'total');
  process.exit(exitCode);
}
